### PR TITLE
21 implementing project information update functionality

### PR DESCRIPTION
**Pull request by k0drin**
### API Endpoint Development
- Developed robust API endpoints to handle GET and POST requests for updating project information.

### Input Validation Implementation
- Implemented back-end validation for the project information input to ensure data accuracy and completeness.

### Real-Time Update Mechanism
- Integrated a real-time update mechanism to ensure that changes are immediately reflected on the startup’s public profile upon saving.

### Revision History Feature
- Developed a backend solution to track and store the history of changes made to the project information.

### Comprehensive Testing
- Conducted extensive testing, including unit tests, integration tests, and user acceptance tests, to ensure the feature's reliability and effectiveness.

WebSocket
SuccessTest py
Daphne

### CODE REVIEW

File: forum/projects/consumers.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        text_data_json = json.loads(text_data)
+        message = text_data_json['message']
+
+        await self.channel_layer.group_send(
+            self.project_group_name,
+            {
+                'type': 'project_update',
+                'message': message
+            }
+        )
```
Comment (mehalyna) [inline forum/projects/consumers.py:67]: Currently, the `receive` method assumes that the incoming data will always contain a message key in the JSON object. You might want to add error handling to catch and handle invalid messages or incorrect JSON data.

File: forum/projects/consumers.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        """
+
+        self.project_id = self.scope['url_route']['kwargs']['project_id']
+        self.project_group_name = f'project_{self.project_id}'
+
+        await self.channel_layer.group_add(
+            self.project_group_name,
+            self.channel_name
+        )
+        await self.accept()
```
Comment (mehalyna) [inline forum/projects/consumers.py:29]: Make sure to validate the users connecting to the WebSocket. You can use Django’s authentication system to ensure that only authorized users are allowed to interact with the WebSocket.

File: forum/projects/serializers.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        Returns:
+            str: The validated project title if unique.
+
+        Raises:
+            serializers.ValidationError: If the project title already exists.
+        """
+
+        if Project.objects.filter(title=value).exists():
+            raise serializers.ValidationError("Project with this title already exists.")
+        return value
```
Comment (mehalyna) [inline forum/projects/serializers.py:43]: Your `validate_title` method checks if a project with the same title already exists. This is fine but it doesn't take the startup into account, meaning if two startups have projects with the same title, it would fail the validation.
If project titles should be unique per startup, include the startup in the validation.

File: forum/projects/serializers.py
Comment (mehalyna) [inline forum/projects/serializers.py:1]: Consider adding a `validate` method to perform validations that depend on multiple fields, like ensuring that the `planned_finish_date` is `after planned_start_date`.

File: forum/projects/signals.py
Comment (mehalyna) [inline forum/projects/signals.py:1]: _Suggestion for future_
If performance becomes an issue with frequent updates, you might want to throttle updates or batch them together using something like Redis or Celery to avoid sending a large number of WebSocket messages.

File: forum/projects/tests.py [hunk 4]
```diff
@@ -1,3 +1,156 @@
...
+        self.management_url = reverse('project-management')
+
+    def test_get_project_history(self):
+        """
+        Tests that retrieving a project's history returns a 404 error if the project does not exist.
+
+        This method sends a GET request to the project history endpoint using a non-existent project ID.
+        """
+        response = self.client.get(self.history_url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```
Comment (mehalyna) [inline forum/projects/tests.py:66]: Consider adding a test where you first create a project and then retrieve its history to ensure both cases are covered (success and failure).

File: forum/projects/tests.py [hunk 5]
```diff
@@ -1,3 +1,156 @@
...
+            'required_amount': '10000.00',
+            'status': ProjectStatus.PLANNED,
+            'planned_start_date': '2024-10-10',
+            'planned_finish_date': '2024-12-10',
+            'media': None
+        }
+
+        response = self.client.post(self.management_url, data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mehalyna) [inline forum/projects/tests.py:94]: After asserting the 201 status code, you might want to add additional assertions to verify that the project data is correctly saved to the database.

File: forum/projects/tests.py [hunk 6]
```diff
@@ -1,3 +1,156 @@
...
+            'description': 'A description for the user project',
+            'required_amount': '5000.00',
+            'status': ProjectStatus.PLANNED,
+            'planned_start_date': '2024-10-10',
+            'planned_finish_date': '2024-12-10'
+        }
+
+        response = self.client.post(self.management_url, data, format='json')
+        print("Response Status Code:", response.status_code)
+        print("Response Data:", response.data)
```
Comment (mehalyna) [inline forum/projects/tests.py:154]: Remove the print statements once you’ve confirmed that everything is working, as they clutter the test output.

File: forum/projects/tests.py [hunk 7]
```diff
@@ -1,3 +1,156 @@
...
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mehalyna) [inline forum/projects/tests.py:156]: Similar to the `test_create_project`, you can add assertions to ensure that the project was created with the expected values.

Comment (mehalyna) [inline forum/projects/tests.py:156]: Consider adding tests for edge cases like:
 - Missing required fields in the POST data (e.g., title, description).
 - Invalid data types (e.g., sending a string for required_amount).
 - Unauthorized users trying to create a project.

File: forum/projects/views.py [hunk 8]
```diff
@@ -1,3 +1,93 @@
...
+        Returns:
+            Response: A response containing the serialized project
+            data with a 201 status code if successful, or error
+            messages with a 400 status code if validation fails.
+        """
+        serializer = self.get_serializer(data=request.data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (mehalyna) [inline forum/projects/views.py:53]: Since you are already using `ListCreateAPIView`, which automatically provides GET and POST methods, you don't need to manually define the post method unless you need custom behavior. DRF's create() method is already handling the logic for you, so this part can be removed for brevity.

File: forum/projects/views.py [hunk 9]
```diff
@@ -1,3 +1,93 @@
...
+
+    Attributes:
+        queryset (QuerySet): A queryset of all historical Project instances.
+        serializer_class (Serializer): The serializer used to serialize
+            Project history data.
+    """
+
+    queryset = Project.history.all()
+    serializer_class = ProjectSerializer
+
```
Comment (mehalyna) [inline forum/projects/views.py:72]: You should add `IsAuthenticated` to ensure that only authenticated users can access the project history. This protects sensitive information about projects.

**Review by mehalyna — CHANGES_REQUESTED** [review]

**Review by mehalyna — APPROVED** [review]

**Review by AnnKvytko — APPROVED** [review]

**Review by YuTumchuk — APPROVED** [review]
